src/hotspot/share/gc/g1/g1HeapVerifier.cpp

```diff
@@ -367,9 +367,17 @@ class VerifyRegionClosure: public HeapRegionClosure {
     guarantee(!r->is_young() || r->rem_set()->is_complete(), "Remembered set for Young region %u must be complete, is %s", r->hrm_index(), r->rem_set()->get_state_str());
     // Humongous and old regions regions might be of any state, so can't check here.
     guarantee(!r->is_free() || !r->rem_set()->is_tracked(), "Remembered set for free region %u must be untracked, is %s", r->hrm_index(), r->rem_set()->get_state_str());
-    // Verify that the continues humongous regions' remembered set state matches the
-    // one from the starts humongous region.
-    if (r->is_continues_humongous()) {
+
+    // For archive regions, verify there are no heap pointers to non-pinned regions.
+    if (r->is_closed_archive()) {
+      VerifyObjectInArchiveRegionClosure verify_oop_pointers(r, false);
+      r->object_iterate(&verify_oop_pointers);
```

Comment:
This should also set `_failures` if there are any; the previous code early exited via `return true`.